### PR TITLE
Feature/issue 706 mdivide right

### PR DESCRIPTION
**Pull request by bob-carpenter**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Fix template parameters in `mdivide_right_tri` so that it could be called with lazy Eigen transpose rather than our copy-heavy version.

#### Intended Effect:

Speedup.

#### How to Verify:

Unit tests.

#### Side Effects:

No.

#### Documentation:

n/a

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Any condition that involves only static variables will be evaluated statically. The ones that evaluate to false will be eliminated.  No branch prediction necessary!  This is one of the main reasons C++ can be so fast.

**Comment by bob-carpenter** [comment]
Jenkins, retest this please.

**Comment by bob-carpenter** [comment]
This looks like another Jenkins timeout.  The previous Travis failure was a timeout.

**Comment by wds15** [comment]
what are your thoughts on applying the same fix to mdivide_right? This pull currently adresses only mdivide_right_tri, but that non-tri function suffers as well from lots of copying. I can push those changes to the branch if you want unless you prefer to keep you submitter and me the reviewer... I don't mind.

**Comment by bob-carpenter** [comment]
Either way---merge the current pull request or keep adding.

I'm happy to keep reviewing.

**Comment by wds15** [comment]
Ok, I made that change. Once that passes the usual test chain we can merge from my perspective.

**Comment by wds15** [comment]
this pull is already on yellow forever. Is the Jenkins stalled?

**Comment by bob-carpenter** [comment]
@seantalts or @syclik know how to diagnose this?  It's already created 25MB of output and it looks like it's working on more, but maybe it's hung in an infinite loop somewhere.

**Comment by bob-carpenter** [comment]
jenkins, retest this please.

**Comment by bob-carpenter** [comment]
No idea how this one's hung up.  I tried everything I could find to kill it and it looks like it's been aborted, but the display is still a yellow circle here.

**Comment by seantalts** [comment]
Yeah, it's been aborted. Maybe it's supposed to leave it as a yellow circle if it's aborted?

To restart, you need to follow the instructions for Retrying a Job here: http://discourse.mc-stan.org/t/new-jenkins-jobs-tutorial/2383

**Comment by wds15** [comment]
I think I followed your instructions correctly, @seantalts. To me it looks like nothing happens. There is no build 3 which has been started. If you could have a look that would be great. Many thanks.

**Comment by bob-carpenter** [comment]
Ditto.  Couldn't restart.  Nothing happens.  I assume this is related to the later out-of-memory on Jenkins message.

**Comment by seantalts** [comment]
Somehow this PR is generating ~45 GB of log files per build - any idea why that would be?

**Comment by wds15** [comment]
No idea. All we did is to remove unneeded copying in the mdivide_right and mdivide_right_tri functions. Things look green now... so we can merge?

File: test/unit/math/fwd/mat/fun/mdivide_right_tri_test.cpp [hunk 1]
```diff
@@ -1,6 +1,20 @@
...
+  using Eigen::MatrixXd;
+  using Eigen::RowVectorXd;
+  using stan::math::mdivide_right_tri;
+
+  MatrixXd b(1, 2);
+  b << 2, 3;
+  MatrixXd A(2, 2);
+  A << 3, 0, 1, 4;
+  MatrixXd b_under_A = mdivide_right_tri<Lower>(b, A);
+  SUCCEED();
```
Comment (syclik) [inline test/unit/math/fwd/mat/fun/mdivide_right_tri_test.cpp]: What's the purpose of this test? It looks like it's just checking that it doesn't crash. If so, look at the death test macros.

**Comment by seantalts** [comment]
(re: long log files - I suspect it was some kind of fluke / bug / one off in Jenkins unrelated to the diff, will investigate further).

**Comment by bob-carpenter** [comment]
I removed the offending test.  I put that there to debug.  If we needed it, it would've been `EXPECT_NO_THROW`.  

On a more general policy note, I was under the impression that we'd moved to a single reviewer per PR to avoid situations where reviewers keep filing in and requesting changes.  The decision arose to favor timely merges and less frustrated developers over potential quality improvements from multiple reviewers.  

There was also a request to never comment on format and to go easy on style.  So I've censored all of my comments that aren't efficiency related.

We can change the policy (or my understanding of what we're trying to do), but we should make it clear what we're trying to do.  I'm explicitly staying away from PR reviewing that requests changes if another reviewer has already started a review.

@syclik has the final say on stan-dev/math, so even if my process understanding is correct, he can overrule the final decisions here.

**Comment by syclik** [comment]
Thanks, @bob-carpenter. Want to move the policy discussion to discourse? 

My review wasn't meant to hinder progress -- I just peeked at it because I was trying to see if I knew why continuous integration was failing. Just saw the test in there.

I don't know what's going on with the Jenkins issue. Haven't gotten comfortable enough to debug it.

**Comment by bob-carpenter** [comment]
I can't keep up with Discourse, but I don't mind if it's going to be a slow discussion.

I don't think anyone sets out with the goal to hinder progress.  It's just that we're each picky in our own way and when we pile on requests for changes, pull requests get bogged down.  This one's now failing tests, which means we have to go kick Jenkins or whatever again.

**Comment by syclik** [comment]
Re: Discourse. I hear you. I'm having trouble keeping up these days. I
didn't want this pull request to be the place for this discussion and it be
buried after we're done.

Reviews are they aren't all at the same level of detail / depth. We don't
have a policy or even guidelines on what to look for within a pull request.
That'll probably need to go hand-in-hand with the policy on reviewers.

Kicking Jenkins... I know @seantalts put up directions on Discourse. I need
to internalize them soon. I'm not sure how to do it off the top of my head.

**Comment by wds15** [comment]
Kicking Jenkins earlier today gave a happy green here. Merge?

**Comment by bob-carpenter** [comment]
That's 16 tests (tri/full, var/double, var/double, matrix/vector).  I started down this road and each one's so complicated I backed into writing a metaprogramming framework.  Two functors and several traits metaprograms later, I gave up for now.  This one's going to have to wait until I have time to do this properly or someone with more copy-and-paste stamina than me does it by brute force.

**Comment by wds15** [comment]
The changes in this pull lead to significant speedups for these functions. Can we file an issue for additional test, but merge already this?

**Comment by syclik** [comment]
Bob, can you commit at least one your tests on the branch? Perhaps we can split the workload and it'll go much faster.

**Comment by syclik** [comment]
> The changes in this pull lead to significant speedups for these functions. Can we file an issue for additional test, but merge already this?

@wds15, no. 

I understand what you're asking for, but the tests (and doc) should come with the pull request. We tried to do what you suggested in the past and it hasn't worked out in many cases.

I think @bob-carpenter is trying to do more than what I envisioned? But I can't tell. (I probably should have been much more explicit on what I wanted tested. I didn't want it fully tested, I just wanted to know that it wouldn't crash when it was instantiated in the 16 ways. We already don't have tests for correct behavior, but I'm not comfortable rolling out code that affects so many things without having at least some test that touches the code once.)

**Comment by bob-carpenter** [comment]
It's not a new function and there are not tests now, but I agree that all of our functions should be tested.  The problem is that's effectively a stop-work order on autodiff given the current lack of a testing framework.  The combinatorics are just too prohibitive in these heavily templated matrix functions.  This square function requires dozens of tests to be exhaustive, but there are only two shapes.  Same thing for the triangular version.

These particular functions are super painful to work with because of the six template parameters and complicated traits return types that depend on the shape of the input.  I never got anyting working, but here's where I was going:

```
template <typename T1, typename T2, int R1, int C2>
struct mdivide_right_return {
  using boost::math::tools::promote_args;
  typedef Eigen::Matrix<typename promote_args<T1, T2>::type, R1, C2>
  type;
};

struct mdiv_fun {
  template <typename T1, typename T2, int R1, int C1, int R2, int C2>
  inline typename mdivide_right_return<T1, T2, R1, C2>::type
  operator()(const Eigen::Matrix<T1, R1, C1> &A,
             const Eigen::Matrix<T2, R2, C2> &B,
             int i) const {
    return mdivide_right(A, B);
  }
};

struct expanded_fun {
  template <typename T1, typename T2, int R1, int C1, int R2, int C2>
  inline typename mdivide_right_return<T1, T2, R1, C2>::type
  operator()(const Eigen::Matrix<T1, R1, C1> &A,
             const Eigen::Matrix<T2, R2, C2> &B,
             int i) const {
    return multiply(A, inverse(B))(i);
  }
};

template <typename T, int R, int C>
Eigen::Matrix<stan::math::var, R, C>
to_var(const Eigen::Matrix<T, R, C>& A) {
  Eigen::Matrix<stan::math::var, R, C> Avar(A.rows(), A.cols());
  for (int i = 0; i < A.size(); ++i)
    Avar(i) = A(i);
  return Avar;
}

template <typename T1, typename T2, int R1, int C1, int R2, int C2>
void same_derivs(const Eigen::Matrix<T1, R1, C1> &A,
                 const Eigen::Matrix<T2, R2, C2> &B) const {
  mdiv_fun f1;
  expanded_fun f2;
  int size = A.rows() * B.rows();
  for (int i = 0; i < size; ++i) {
    var y1 = f1(A, B)(i);
    std::vector<double> grad1;
    y1.grad(to_vector(A, B), g);
  }


    mdivide_right_return<T1, T2, R1, C2>::type C2 = mdivide_right(A, B);
    std::vector<double> grad2;
}




template <int R1, int C1, int R2, int C2>
void test_mdivide_right_derivs(const Eigen::Matrix<double, R1, C1> &A,
                               const Eigen::Matrix<double, R2, C2> &B) const {
  using stan::math::var;
  // var, var
  Eigen::Matrix<var, R1, C1> Av1 = to_var(A);
  Eigen::Matrix<var, R2, C2> Av2 = to_var(B);
  validate_same_derivs(Av1, Av2);

}

TEST(AgradRevMatrix, mdivide_right_derivs_mm_vv) {
  using stan::math::matrix_d;
  using stan::math::matrix_v;
  using stan::math::mdivide_right;

  matrix_v A1(2, 2);
  A1 << 1, -2, 3, 17;
  matrix_v B1(2, 2);
  B1 << 3, 4, 1, 9;
  matrix_v C1 = mdivide_right(A1, B1);

  matrix_v A2(2, 2);
  A2 << 1, -2, 3, 17;
  matrix_v B2(2, 2);
  B2 << 3, 4, 1, 9;
  matrix_v C2 = multiply(A2, inverse(B2));

  validate_same_derivs(C1, C2);


}
```

**Comment by bob-carpenter** [comment]
This case really needs 76 tests just for autodiff behavior and 4 for scalar behavior (I'm assuming the latter are in place already, but maybe not).  

I'm going to start by documenting the testing framework I built for scalar autodiff and then sketching out  a test for multivariates.  I hope I can get to that after taking on the `map` function for MPI.  I'm OK putting this optimization on hold until then.  To test this properly, we need a whole bunch of test cases:

* mdivide_right/mdivide_right_tri
* autodiff/double on inverted argument
* autodiff/double on non-inverted argument
* vector/matrix on non-inverted argument

times the autodif type (which is only a single choice) for each combo of the above:

* `var`, `fvar<double>`, `fvar<fvar<double>>`, `fvar<var>`, and `fvar<fvar<var>>`

I count 80 tests.  I don't want to do these by hand.  It's just not sustainable.

I'm OK waiting to stage this (and all subequent) performance improvements until after someone has time to implement an autodiff test framework.  I'd ideally like to work on finishing tests for what we have before writing new code.

**Comment by seantalts** [comment]
Hey @syclik, it came up in the meeting that we didn't know exactly what you meant for the tests here. Can you give more of a spec so we know what we're aiming at, if we really need to make an automated testing framework for this?

**Comment by bob-carpenter** [comment]
The testing framework isn't necessary in the sense that other functions have this all done by hand.  

It would be most useful to test the derivatives for all combinations of inputs at all orders.  I wouldn't want to do that by hand for this pair of functions.

**Comment by syclik** [comment]
@seantalts, thanks for asking for more clarity. @bob-carpenter, I don't see the ability to have both vector and matrix on the non-inverted type. So the total count should be 40, but I was really just asking for 8 tests. I was suggesting a few quick tests for `var`, but not every instantiation of higher-order autodiff. More below.

** What I'm worried about **

The thing that scares me about this pull request is the use of Eigen's template expressions in this complicated way. It scares me because it may affect our autodiff stack in some unexpected way.

I'm afraid that under some instantiation, the code may seg fault, allocate memory on the wrong stack, or does something else that's wild and beyond our normal ability to imagine it.

Fortunately, I think that concern can be fixed with a simple test or two.

** What I want tested **

1. Instantiations in the things we care about now: `double` and `var` for the templated arguments.
2. Check that calling the function and grad for each of the instantiations doesn't crash.
3. Check the autodiff stack before and after the grad call.

I think this will help with maintenance (just in case it works for current Eigen, but not future Eigen). It should also help when anyone wants to push forward with specialized derivatives for this function.

** What I'm not worried about **

Funny enough, I'm less worried about correctness. We've gone on this far without having tests for the function value or the gradients for this function.

Of course, I'd like to have it, but I don't think it's on the person that changes a function to have to carry that burden. It's technical debt. I'm ok carrying that technical debt forward. I'm not ok if a function like this starts "leaking" memory or seg faults on a user because we didn't call each instantiation once.

I'm also less worried about the instantiations we're not calling from Stan now. We will eventually need to test all of `fvar`, but that's not a burden I'm putting on this pull request.


** More specifically **

I think we need to do two things for this pull request:
1. modify `stan/math/rev/mat/mdivide_right_test.cpp` to call `.grad()` on any element just to make sure the gradient function is being called. (Yes, it doesn't do that now!)
2. copy that test to `mdivide_right_tri_test.cpp` and instantiate that in the same way.

Of course, I'd like to see all the gradients tested through, but I'm not asking for it (because we haven't tested it in the past and you're just making a modification to this function).

**Comment by bob-carpenter** [comment]
For

mdivide_right(a, B) = a * inv(B)

a can be a row vector or a matrix.

The types should probably be simplified because B has to
be a matrix.

> On Feb 9, 2018, at 2:38 PM, Daniel Lee <notifications@github.com> wrote:
> 
> @seantalts, thanks for asking for more clarity. @bob-carpenter, I don't see the ability to have both vector and matrix on the non-inverted type. So the total count should be 40, but I was really just asking for 8 tests. I was suggesting a few quick tests for var, but not every instantiation of higher-order autodiff. More below.
> 
> ** What I'm worried about **
> 
> The thing that scares me about this pull request is the use of Eigen's template expressions in this complicated way. It scares me because it may affect our autodiff stack in some unexpected way.

Only arithmetic affects the stack.

> I'm afraid that under some instantiation, the code may seg fault, allocate memory on the wrong stack, or does something else that's wild and beyond our normal ability to imagine it.

That could only happen if some piece of our arithmetic
did that.  That would be a pretty deep bug, but this
isn't the place to test that.


> Fortunately, I think that concern can be fixed with a simple test or two.
> 
> ** What I want tested **
> 
> 	• Instantiations in the things we care about now: double and var for the templated arguments.
> 	• Check that calling the function and grad for each of the instantiations doesn't crash.
> 	• Check the autodiff stack before and after the grad call.

What did you want to check the stack for?   If this is something
you're going to want in general, we should add it to the existing
autodiff tests.

The stack before and after the grad() call is really grad's responsibility.
Are you worried about nested calls somehow accidentally getting
called or an error in our calling of chain() somehow?

> I think this will help with maintenance (just in case it works for current Eigen, but not future Eigen). It should also help when anyone wants to push forward with specialized derivatives for this function.
> 
> ** What I'm not worried about **
> 
> Funny enough, I'm less worried about correctness. We've gone on this far without having tests for the function value or the gradients for this function.
> 
> Of course, I'd like to have it, but I don't think it's on the person that changes a function to have to carry that burden. It's technical debt. I'm ok carrying that technical debt forward. I'm not ok if a function like this starts "leaking" memory or seg faults on a user because we didn't call each instantiation once.

I'm much more worried about this because I can't imagine any
of the ways anything you're worried about above could go wrong,
since they'd have to go wrong in our underlying arithmetic.

> I'm also less worried about the instantiations we're not calling from Stan now. We will eventually need to test all of fvar, but that's not a burden I'm putting on this pull request.

OK.  What about with new functions?


> ** More specifically **
> 
> I think we need to do two things for this pull request:
> 
> 	• modify stan/math/rev/mat/mdivide_right_test.cpp to call .grad() on any element just to make sure the gradient function is being called. (Yes, it doesn't do that now!)

I completely agree the current tests are insufficient.

> 	• copy that test to mdivide_right_tri_test.cpp and instantiate that in the same way.

But with triangular matrices.

> Of course, I'd like to see all the gradients tested through, but I'm not asking for it (because we haven't tested it in the past and you're just making a modification to this function).

I'm not clear on this distinction.  I'd be more worried that a change
will introduce an error in calculating the value than I would be
on it finding a bug buried in our arithmetic somewhere.

- Bob

**Comment by syclik** [comment]
> Only arithmetic affects the stack.

Yes, but we're relying on Eigen here to do the arithmetic.

> That could only happen if some piece of our arithmetic did that.  That would be a pretty deep bug, but this isn't the place to test that.

Yes and no. Let me elaborate a bit.

We're relying on Eigen for these operations. We've already seen that different versions of Eigen produce different results (we have tests conditional on version of Eigen in our codebase). We've run into Eigen bugs, compiler bugs, and even different behavior based on compiler. Let alone cross-platform incompatibility.

The right place to test would possibly be some sort of integration test that verifies that every Eigen operation works with our autodiff variables and stack. Then we could just rely on that behavior to move forward. We don't have that and I really don't want to test against Eigen for everything that Eigen does.

Instead, I think this is one place we can test that the behavior is correct, including how it plays out on the autodiff stack. This pull request is now using Eigen's functions instead of what we've tested all the way down (or should have tested). My fear here is that a future version of Eigen will do something different and we won't know. 

The reasonable thing may be to ignore that fear. Especially if you're convinced that Eigen's implementation can't allocate things on the wrong stack and everything on the stack will be ok.

> Are you worried about nested calls somehow accidentally getting called or an error in our calling of chain() somehow?

Not worried about nested calls. More worried that certain things don't end up on our stack and some place else.

> I'm much more worried about this because I can't imagine any of the ways anything you're worried about above could go wrong, since they'd have to go wrong in our underlying arithmetic.

Ok! We should worry about correctness. Especially if you're convinced that none of the things that I'm worried about could go wrong.

> OK.  What about with new functions?

We'll discuss on discourse instead of this PR.

**Comment by syclik** [comment]
And you're absolutely right. It's 80, not 40. It's hard to keep these instantiations correct.

I'm going to approve this pull request. I think we know where the weaknesses are. It's not new technical debt we're carrying forward. Can I ask you to add a new issue to request testing for this particular function?

**Comment by bob-carpenter** [comment]
Issue for testing: https://github.com/stan-dev/math/issues/763

**Comment by wds15** [comment]
So we can merge?

**Pull request merged**